File: apps/web/tests/lib/streaming.test.ts
import { describe, it, expect } from "vitest";
import {
  parseSSE,
  parseOpenAIEvent,
  parseAnthropicEvent,
  parseInternalEvent,
  StreamAccumulator,
  detectFormat,
} from "@/lib/streaming";

describe("parseSSE", () => {
  it("parses simple data events", () => {
    const text = 'data: {"hello":"world"}\n\n';
    const events = [...parseSSE(text)];
    expect(events).toHaveLength(1);
    expect(events[0].data).toBe('{"hello":"world"}');
    expect(events[0].event).toBeUndefined();
  });

  it("parses named events", () => {
    const text = 'event: message_start\ndata: {"type":"message_start"}\n\n';
    const events = [...parseSSE(text)];
    expect(events).toHaveLength(1);
    expect(events[0].event).toBe("message_start");
  });

  it("parses multiple events", () => {
    const text =
      'event: chunk\ndata: {"a":1}\n\nevent: chunk\ndata: {"b":2}\n\n';
    const events = [...parseSSE(text)];
    expect(events).toHaveLength(2);
  });

  it("ignores comments", () => {
    const text = ': keepalive\ndata: {"x":1}\n\n';
    const events = [...parseSSE(text)];
    expect(events).toHaveLength(1);
    expect(events[0].data).toBe('{"x":1}');
  });
});

describe("parseOpenAIEvent", () => {
  it("parses content delta", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "gpt-4o",
        choices: [{ index: 0, delta: { content: "Hello" }, finish_reason: null }],
      }),
    };
    const events = parseOpenAIEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("content");
    expect(events[0].content).toBe("Hello");
  });

  it("parses role delta", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "gpt-4o",
        choices: [{ index: 0, delta: { role: "assistant" }, finish_reason: null }],
      }),
    };
    const events = parseOpenAIEvent(evt);
    // Role-only delta produces no events (no content/thinking/toolcall/finish)
    expect(events).toHaveLength(0);
  });

  it("parses finish event", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "gpt-4o",
        choices: [{ index: 0, delta: {}, finish_reason: "stop" }],
      }),
    };
    const events = parseOpenAIEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("finish");
    expect(events[0].finishReason).toBe("stop");
  });

  it("parses [DONE]", () => {
    const events = parseOpenAIEvent({ data: "[DONE]" });
    expect(events).toHaveLength(0);
  });

  it("parses usage chunk", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "gpt-4o",
        choices: [],
        usage: { prompt_tokens: 10, completion_tokens: 5, total_tokens: 15 },
      }),
    };
    const events = parseOpenAIEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("usage");
    expect(events[0].usage?.totalTokens).toBe(15);
  });

  it("parses tool call delta", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "gpt-4o",
        choices: [
          {
            index: 0,
            delta: {
              tool_calls: [
                {
                  index: 0,
                  id: "call_1",
                  type: "function",
                  function: { name: "get_weather", arguments: '{"loc' },
                },
              ],
            },
            finish_reason: null,
          },
        ],
      }),
    };
    const events = parseOpenAIEvent(evt);
    expect(events).toHaveLength(2);
    expect(events[0].type).toBe("tool_call_start");
    expect(events[0].toolCall?.id).toBe("call_1");
    expect(events[1].type).toBe("tool_call_delta");
    expect(events[1].toolCall?.arguments).toBe('{"loc');
  });

  it("parses reasoning_content", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "o1-preview",
        choices: [{ index: 0, delta: { reasoning_content: "thinking..." }, finish_reason: null }],
      }),
    };
    const events = parseOpenAIEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("thinking");
    expect(events[0].thinking).toBe("thinking...");
  });

  it("parses mid-stream error", () => {
    const evt = {
      data: JSON.stringify({
        id: "cmpl-1",
        object: "chat.completion.chunk",
        created: 1000,
        model: "gpt-4o",
        error: { code: "server_error", message: "disconnected" },
        choices: [{ index: 0, delta: {}, finish_reason: "error" }],
      }),
    };
    const events = parseOpenAIEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("error");
    expect(events[0].error?.code).toBe("server_error");
  });
});

describe("parseAnthropicEvent", () => {
  it("parses content_block_delta text", () => {
    const evt = {
      event: "content_block_delta",
      data: JSON.stringify({
        type: "content_block_delta",
        index: 0,
        delta: { type: "text_delta", text: "Hello" },
      }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("content");
    expect(events[0].content).toBe("Hello");
  });

  it("parses thinking delta", () => {
    const evt = {
      event: "content_block_delta",
      data: JSON.stringify({
        type: "content_block_delta",
        index: 0,
        delta: { type: "thinking_delta", thinking: "reasoning..." },
      }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("thinking");
    expect(events[0].thinking).toBe("reasoning...");
  });

  it("parses tool_use content_block_start", () => {
    const evt = {
      event: "content_block_start",
      data: JSON.stringify({
        type: "content_block_start",
        index: 1,
        content_block: { type: "tool_use", id: "tu_1", name: "get_weather", input: {} },
      }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("tool_call_start");
    expect(events[0].toolCall?.id).toBe("tu_1");
    expect(events[0].toolCall?.name).toBe("get_weather");
  });

  it("parses input_json_delta", () => {
    const evt = {
      event: "content_block_delta",
      data: JSON.stringify({
        type: "content_block_delta",
        index: 1,
        delta: { type: "input_json_delta", partial_json: '{"location":' },
      }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("tool_call_delta");
    expect(events[0].toolCall?.arguments).toBe('{"location":');
  });

  it("parses message_delta stop reason", () => {
    const evt = {
      event: "message_delta",
      data: JSON.stringify({
        type: "message_delta",
        delta: { stop_reason: "end_turn" },
        usage: { output_tokens: 15 },
      }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(2);
    expect(events[0].type).toBe("finish");
    expect(events[0].finishReason).toBe("end_turn");
    expect(events[1].type).toBe("usage");
  });

  it("parses error event", () => {
    const evt = {
      event: "error",
      data: JSON.stringify({
        type: "error",
        error: { type: "overloaded_error", message: "Overloaded" },
      }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("error");
    expect(events[0].error?.code).toBe("overloaded_error");
  });

  it("ignores ping events", () => {
    const evt = {
      event: "ping",
      data: JSON.stringify({ type: "ping" }),
    };
    const events = parseAnthropicEvent(evt);
    expect(events).toHaveLength(0);
  });
});

describe("parseInternalEvent", () => {
  it("parses chunk with content", () => {
    const evt = {
      event: "chunk",
      data: JSON.stringify({ delta: { content: "Hello" } }),
    };
    const events = parseInternalEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("content");
  });

  it("parses thinking event", () => {
    const evt = {
      event: "thinking",
      data: JSON.stringify({ content: "reasoning..." }),
    };
    const events = parseInternalEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("thinking");
  });

  it("parses finish event", () => {
    const evt = {
      event: "finish",
      data: JSON.stringify({ reason: "stop", usage: { promptTokens: 10 } }),
    };
    const events = parseInternalEvent(evt);
    expect(events).toHaveLength(1);
    expect(events[0].type).toBe("finish");
  });
});

describe("StreamAccumulator", () => {
  it("accumulates content", () => {
    const acc = new StreamAccumulator();
    acc.addEvent({ type: "content", content: "Hello" });
    acc.addEvent({ type: "content", content: " world" });
    expect(acc.content).toBe("Hello world");
  });

  it("accumulates thinking", () => {
    const acc = new StreamAccumulator();
    acc.addEvent({ type: "thinking", thinking: "let me think" });
    expect(acc.thinking).toBe("let me think");
  });

  it("accumulates tool calls", () => {
    const acc = new StreamAccumulator();
    acc.addEvent({ type: "tool_call_start", toolCall: { id: "c1", name: "calc" } });
    acc.addEvent({ type: "tool_call_delta", toolCall: { arguments: '{"x":' } });
    acc.addEvent({ type: "tool_call_delta", toolCall: { arguments: "1}" } });
    const msg = acc.message();
    expect(msg.toolCalls).toHaveLength(1);
    expect(msg.toolCalls[0].id).toBe("c1");
    expect(msg.toolCalls[0].name).toBe("calc");
    expect(msg.toolCalls[0].arguments).toBe('{"x":1}');
  });

  it("captures finish reason and usage", () => {
    const acc = new StreamAccumulator();
    acc.addEvent({ type: "finish", finishReason: "stop" });
    acc.addEvent({
      type: "usage",
      usage: { promptTokens: 10, completionTokens: 5, totalTokens: 15 },
    });
    const msg = acc.message();
    expect(msg.finishReason).toBe("stop");
    expect(msg.usage?.totalTokens).toBe(15);
  });

  it("returns copy of tool calls", () => {
    const acc = new StreamAccumulator();
    acc.addEvent({ type: "tool_call_start", toolCall: { id: "c1", name: "f" } });
    const calls1 = acc.toolCalls;
    const calls2 = acc.toolCalls;
    expect(calls1).not.toBe(calls2); // different array references
    expect(calls1).toEqual(calls2);
  });
});

describe("detectFormat", () => {
  it("detects OpenAI format", () => {
    const events = [
      { data: '{"object":"chat.completion.chunk","choices":[],"model":"gpt-4o"}' },
    ];
    expect(detectFormat(events)).toBe("openai");
  });

  it("detects Anthropic format", () => {
    const events = [{ event: "message_start", data: '{"type":"message_start"}' }];
    expect(detectFormat(events)).toBe("anthropic");
  });

  it("detects internal format", () => {
    const events = [{ event: "chunk", data: '{"delta":{"content":"hi"}}' }];
    expect(detectFormat(events)).toBe("internal");
  });

  it("defaults to OpenAI", () => {
    const events = [{ data: "malformed" }];
    expect(detectFormat(events)).toBe("openai");
  });
});

File: .claude/worktrees/security-bugfixes-2026-06-02/apps/backend/pkg/llm/streaming/streaming_test.go
package streaming

import (
	"bytes"
	"context"
	"encoding/json"
	"strings"
	"testing"

	"dra-platform/backend/pkg/llm"
)

// --- Accumulator tests ---

func TestAccumulator_Content(t *testing.T) {
	acc := NewAccumulator()
	acc.Add(llm.StreamChunk{Delta: llm.Message{Role: llm.RoleAssistant, Content: "Hello"}})
	acc.Add(llm.StreamChunk{Delta: llm.Message{Content: " world"}})

	msg, done := acc.Message()
	if done {
		t.Error("expected not done")
	}
	if msg.Content != "Hello world" {
		t.Errorf("content = %q, want %q", msg.Content, "Hello world")
	}
	if msg.Role != llm.RoleAssistant {
		t.Errorf("role = %q", msg.Role)
	}
}

func TestAccumulator_ToolCalls(t *testing.T) {
	acc := NewAccumulator()
	acc.Add(llm.StreamChunk{
		Delta: llm.Message{
			Role: llm.RoleAssistant,
			ToolCalls: []llm.ToolCall{
				{ID: "call_1", Type: "function", Function: llm.ToolCallFunction{Name: "calc", Arguments: json.RawMessage(`{"ex`)}},
			},
		},
	})
	acc.Add(llm.StreamChunk{
		Delta: llm.Message{
			ToolCalls: []llm.ToolCall{
				{Function: llm.ToolCallFunction{Arguments: json.RawMessage(`pression":"1+1"}`)}},
			},
		},
	})

	msg, done := acc.Message()
	if done {
		t.Error("expected not done")
	}
	if len(msg.ToolCalls) != 1 {
		t.Fatalf("tool calls = %d, want 1", len(msg.ToolCalls))
	}
	if string(msg.ToolCalls[0].Function.Arguments) != `{"expression":"1+1"}` {
		t.Errorf("arguments = %q", string(msg.ToolCalls[0].Function.Arguments))
	}
}

func TestAccumulator_FinishReason(t *testing.T) {
	acc := NewAccumulator()
	acc.Add(llm.StreamChunk{
		Delta:        llm.Message{Role: llm.RoleAssistant},
		FinishReason: ptrFinishReason(llm.FinishReasonStop),
	})

	msg, done := acc.Message()
	if !done {
		t.Error("expected done")
	}
	if msg.Role != llm.RoleAssistant {
		t.Errorf("role = %q", msg.Role)
	}
}

func TestAccumulator_Thinking(t *testing.T) {
	acc := NewAccumulator()
	acc.Add(llm.StreamChunk{Thinking: "let me think"})
	acc.Add(llm.StreamChunk{Delta: llm.Message{Content: "answer"}})

	msg, _ := acc.Message()
	if len(msg.ContentBlocks) != 1 {
		t.Fatalf("content blocks = %d, want 1", len(msg.ContentBlocks))
	}
	if msg.ContentBlocks[0].Type != llm.ContentTypeThinking {
		t.Errorf("block type = %q", msg.ContentBlocks[0].Type)
	}
	if msg.ContentBlocks[0].Thinking != "let me think" {
		t.Errorf("thinking = %q", msg.ContentBlocks[0].Thinking)
	}
}

func TestAccumulator_Empty(t *testing.T) {
	acc := NewAccumulator()
	msg, done := acc.Message()
	if done {
		t.Error("expected not done")
	}
	if msg.Content != "" {
		t.Errorf("content = %q", msg.Content)
	}
}

// --- OpenAI Writer tests ---

func TestOpenAIStreamWriter_BasicChunk(t *testing.T) {
	var buf bytes.Buffer
	w := NewOpenAIStreamWriter(&buf, "gpt-4o")

	chunk := &llm.StreamChunk{
		Index: 0,
		Delta: llm.Message{Role: llm.RoleAssistant, Content: "Hello"},
	}
	if err := w.WriteChunk(chunk); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "data: ") {
		t.Error("missing data prefix")
	}
	if !strings.Contains(output, "chat.completion.chunk") {
		t.Error("missing object type")
	}
	if !strings.Contains(output, "Hello") {
		t.Error("missing content")
	}
}

func TestOpenAIStreamWriter_FinishWithUsage(t *testing.T) {
	var buf bytes.Buffer
	w := NewOpenAIStreamWriter(&buf, "gpt-4o")

	usage := &llm.Usage{PromptTokens: 10, CompletionTokens: 20, TotalTokens: 30}
	if err := w.WriteFinish(llm.FinishReasonStop, usage); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "[DONE]") {
		t.Error("missing [DONE]")
	}
	if !strings.Contains(output, "total_tokens") {
		t.Error("missing usage")
	}
}

func TestOpenAIStreamWriter_ToolCall(t *testing.T) {
	var buf bytes.Buffer
	w := NewOpenAIStreamWriter(&buf, "gpt-4o")

	tc := &llm.ToolCall{ID: "call_1", Type: "function", Function: llm.ToolCallFunction{Name: "get_weather"}}
	if err := w.WriteToolCallStart(tc); err != nil {
		t.Fatal(err)
	}
	if err := w.WriteToolCallDelta(0, `{"location":"SF"}`); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "call_1") {
		t.Error("missing tool call ID")
	}
	if !strings.Contains(output, "get_weather") {
		t.Error("missing function name")
	}
	if !strings.Contains(output, "location") || !strings.Contains(output, "SF") {
		t.Error("missing arguments")
	}
}

func TestOpenAIStreamWriter_Thinking(t *testing.T) {
	var buf bytes.Buffer
	w := NewOpenAIStreamWriter(&buf, "o1-preview")

	if err := w.WriteThinking("reasoning..."); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "reasoning_content") {
		t.Error("missing reasoning_content")
	}
	if !strings.Contains(output, "reasoning...") {
		t.Error("missing thinking content")
	}
}

func TestOpenAIStreamWriter_Error(t *testing.T) {
	var buf bytes.Buffer
	w := NewOpenAIStreamWriter(&buf, "gpt-4o")

	if err := w.WriteError("server_error", "provider disconnected"); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "server_error") {
		t.Error("missing error code")
	}
	if !strings.Contains(output, "provider disconnected") {
		t.Error("missing error message")
	}
}

// --- Anthropic Writer tests ---

func TestAnthropicStreamWriter_BasicChunk(t *testing.T) {
	var buf bytes.Buffer
	w := NewAnthropicStreamWriter(&buf, "claude-opus-4")

	chunk := &llm.StreamChunk{
		ID:    "msg_test",
		Index: 0,
		Delta: llm.Message{Content: "Hello"},
	}
	if err := w.WriteChunk(chunk); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "event: message_start") {
		t.Error("missing message_start event")
	}
	if !strings.Contains(output, "event: content_block_start") {
		t.Error("missing content_block_start event")
	}
	if !strings.Contains(output, "event: content_block_delta") {
		t.Error("missing content_block_delta event")
	}
	if !strings.Contains(output, "text_delta") {
		t.Error("missing text_delta type")
	}
}

func TestAnthropicStreamWriter_ToolCall(t *testing.T) {
	var buf bytes.Buffer
	w := NewAnthropicStreamWriter(&buf, "claude-opus-4")

	// First write some content to create a text block
	w.WriteChunk(&llm.StreamChunk{Delta: llm.Message{Content: "Let me check"}})

	tc := &llm.ToolCall{ID: "tu_1", Type: "function", Function: llm.ToolCallFunction{Name: "get_weather"}}
	if err := w.WriteToolCallStart(tc); err != nil {
		t.Fatal(err)
	}
	if err := w.WriteToolCallDelta(0, `{"location":"NYC"}`); err != nil {
		t.Fatal(err)
	}
	if err := w.WriteToolCallEnd(0); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "content_block_stop") {
		t.Error("missing content_block_stop to close text block")
	}
	if !strings.Contains(output, "tool_use") {
		t.Error("missing tool_use content block")
	}
	if !strings.Contains(output, "tu_1") {
		t.Error("missing tool use ID")
	}
	if !strings.Contains(output, "input_json_delta") {
		t.Error("missing input_json_delta")
	}
}

func TestAnthropicStreamWriter_Thinking(t *testing.T) {
	var buf bytes.Buffer
	w := NewAnthropicStreamWriter(&buf, "claude-opus-4")

	if err := w.WriteThinking("deep thought"); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "thinking_delta") {
		t.Error("missing thinking_delta type")
	}
	if !strings.Contains(output, "deep thought") {
		t.Error("missing thinking content")
	}
}

func TestAnthropicStreamWriter_Finish(t *testing.T) {
	var buf bytes.Buffer
	w := NewAnthropicStreamWriter(&buf, "claude-opus-4")

	// Write some content first
	w.WriteChunk(&llm.StreamChunk{Delta: llm.Message{Content: "done"}})

	usage := &llm.Usage{PromptTokens: 10, CompletionTokens: 5}
	if err := w.WriteFinish(llm.FinishReasonStop, usage); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "message_delta") {
		t.Error("missing message_delta event")
	}
	if !strings.Contains(output, "message_stop") {
		t.Error("missing message_stop event")
	}
	if !strings.Contains(output, "end_turn") {
		t.Error("missing end_turn stop reason")
	}
}

func TestAnthropicStreamWriter_Error(t *testing.T) {
	var buf bytes.Buffer
	w := NewAnthropicStreamWriter(&buf, "claude-opus-4")

	if err := w.WriteError("overloaded_error", "Overloaded"); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "event: error") {
		t.Error("missing error event")
	}
	if !strings.Contains(output, "overloaded_error") {
		t.Error("missing error type")
	}
}

// --- Internal Writer tests ---

func TestInternalStreamWriter_BasicChunk(t *testing.T) {
	var buf bytes.Buffer
	w := NewInternalStreamWriter(&buf)

	chunk := &llm.StreamChunk{
		Index: 0,
		Delta: llm.Message{Content: "Hello"},
	}
	if err := w.WriteChunk(chunk); err != nil {
		t.Fatal(err)
	}

	output := buf.String()
	if !strings.Contains(output, "event: chunk") {
		t.Error("missing chunk event")
	}
}

// --- StreamPump tests ---

func TestStreamPump_OpenAI(t *testing.T) {
	var buf bytes.Buffer
	writer := NewOpenAIStreamWriter(&buf, "gpt-4o")
	pump := NewStreamPump(writer)

	ch := make(chan llm.StreamChunk, 3)
	ch <- llm.StreamChunk{Delta: llm.Message{Role: llm.RoleAssistant, Content: "Hi"}}
	ch <- llm.StreamChunk{Delta: llm.Message{Content: " there"}}
	fr := llm.FinishReasonStop
	ch <- llm.StreamChunk{FinishReason: &fr, Usage: &llm.Usage{PromptTokens: 5, CompletionTokens: 2, TotalTokens: 7}}
	close(ch)

	msg, usage, err := pump.Pump(context.Background(), ch)
	if err != nil {
		t.Fatal(err)
	}
	if msg.Content != "Hi there" {
		t.Errorf("content = %q", msg.Content)
	}
	if usage == nil || usage.TotalTokens != 7 {
		t.Error("usage not captured")
	}

	output := buf.String()
	if !strings.Contains(output, "[DONE]") {
		t.Error("missing [DONE]")
	}
}

func TestStreamPump_Anthropic(t *testing.T) {
	var buf bytes.Buffer
	writer := NewAnthropicStreamWriter(&buf, "claude-opus-4")
	pump := NewStreamPump(writer)

	ch := make(chan llm.StreamChunk, 2)
	ch <- llm.StreamChunk{ID: "msg_1", Delta: llm.Message{Content: "Hello"}}
	fr := llm.FinishReasonEndTurn
	ch <- llm.StreamChunk{FinishReason: &fr}
	close(ch)

	msg, _, err := pump.Pump(context.Background(), ch)
	if err != nil {
		t.Fatal(err)
	}
	if msg.Content != "Hello" {
		t.Errorf("content = %q", msg.Content)
	}

	output := buf.String()
	if !strings.Contains(output, "message_start") {
		t.Error("missing message_start")
	}
	if !strings.Contains(output, "message_stop") {
		t.Error("missing message_stop")
	}
}

func TestStreamPump_EmptyStream(t *testing.T) {
	var buf bytes.Buffer
	writer := NewOpenAIStreamWriter(&buf, "gpt-4o")
	pump := NewStreamPump(writer)

	ch := make(chan llm.StreamChunk)
	close(ch)

	msg, _, err := pump.Pump(context.Background(), ch)
	if err != nil {
		t.Fatal(err)
	}
	if msg.Content != "" {
		t.Errorf("content = %q", msg.Content)
	}
}

func TestStreamPump_ContextCancelled(t *testing.T) {
	var buf bytes.Buffer
	writer := NewOpenAIStreamWriter(&buf, "gpt-4o")
	pump := NewStreamPump(writer)

	ch := make(chan llm.StreamChunk) // never closed
	ctx, cancel := context.WithCancel(context.Background())
	cancel() // cancel immediately

	_, _, err := pump.Pump(ctx, ch)
	if err != context.Canceled {
		t.Errorf("err = %v, want Canceled", err)
	}
}

// --- Format helper ---

func TestNewWriterFromIO(t *testing.T) {
	var buf bytes.Buffer

	w := NewWriterFromIO(&buf, FormatOpenAI, "gpt-4o")
	if w.Format() != FormatOpenAI {
		t.Error("expected OpenAI format")
	}

	w = NewWriterFromIO(&buf, FormatAnthropic, "claude-opus-4")
	if w.Format() != FormatAnthropic {
		t.Error("expected Anthropic format")
	}

	w = NewWriterFromIO(&buf, FormatInternal, "model")
	if w.Format() != FormatInternal {
		t.Error("expected Internal format")
	}
}

// --- Anthropic stop reason mapping ---

func TestAnthropicStopReasonFromFinish(t *testing.T) {
	tests := []struct {
		input llm.FinishReason
		want  string
	}{
		{llm.FinishReasonStop, "end_turn"},
		{llm.FinishReasonEndTurn, "end_turn"},
		{llm.FinishReasonLength, "max_tokens"},
		{llm.FinishReasonToolCalls, "tool_use"},
		{llm.FinishReasonContentFilter, "content_filter"},
		{llm.FinishReason("weird"), "weird"},
	}
	for _, tt := range tests {
		got := anthropicStopReasonFromFinish(tt.input)
		if got != tt.want {
			t.Errorf("anthropicStopReasonFromFinish(%q) = %q, want %q", tt.input, got, tt.want)
		}
	}
}

func ptrFinishReason(fr llm.FinishReason) *llm.FinishReason {
	return &fr
}
